docs/architecture/adr-082-data-companion-api.md

```diff
@@ -0,0 +1,361 @@
+    // If an error is returned, Tendermint will crash.
+    rpc CommitBlock (CommitBlockRequest) returns (CommitBlockResponse) {}
+}
+
+// CommitBlockRequest contains information about a block that has been committed
+// by the consensus engine.
+message CommitBlockRequest {
+    // The block committed by the consensus engine.
+    Block block = 1;
+
+    // The results from execution of this block.
+    ExecBlockResults block_results = 2;
+}
+
+// ExecBlockResults contains additional information about the execution of a
+// specific block that will not necessarily be included in the block itself.
+// This data is obtained from the FinalizeBlock ABCI response.
+message ExecBlockResults {
+    repeated Event events = 1;
+    repeated ExecTxResult tx_results = 2;
+}
+
+// CommitBlockResponse is either empty or returns an error. Note that returning
+// an error here will cause Tendermint to crash.
+message CommitBlockResponse {
+    oneof error {
+        UnexpectedBlockError unexpected_block_err = 1;
+    }
+}
+
+// UnexpectedBlockError is an error returned by the server when Tendermint sent
+// it a block that is ahead of the block expected by the server.
+message UnexpectedBlockError {
+    // The height of the block expected by the server.
+    int64 expected_height = 1;
+}
+
+//-----------------------------------------------------------------------------
+// The following types are defined as part of ABCI++, and provided here for
+// reference.
+//-----------------------------------------------------------------------------
+
+// Event allows application developers to attach additional information to
+// ResponseFinalizeBlock, ResponseDeliverTx, ExecTxResult
+// Later, transactions may be queried using these events.
+message Event {
+  string                  type       = 1;
+  repeated EventAttribute attributes = 2 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "attributes,omitempty"];
+}
+
+// EventAttribute is a single key-value pair, associated with an event.
+message EventAttribute {
+  string key   = 1;
+  string value = 2;
+  bool   index = 3;  // nondeterministic
+}
+
+// ExecTxResult contains results of executing one individual transaction.
+//
+// * Its structure is equivalent to #ResponseDeliverTx which will be deprecated/deleted
+message ExecTxResult {
+  uint32         code       = 1;
+  bytes          data       = 2;
+  string         log        = 3;  // nondeterministic
+  string         info       = 4;  // nondeterministic
+  int64          gas_wanted = 5;
+  int64          gas_used   = 6;
+  repeated Event events     = 7
+      [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];  // nondeterministic
+  string codespace = 8;
+}
+```
+
+### Request Buffering
+
+In order to ensure reliable delivery, while also catering for intermittent
+faults, we should allow for configurable buffering of a reasonable, yet small,
+number of requests to the companion service. If this buffer fills up, we should
+panic.
```

Comment:
Hmm, I would conceptualize this as backpressure. It doesn't cause unbounded memory growth and it doesn't hang the running goroutines responsible for consensus. Instead, it halts them completely along with all other routines in the program. The only distinction I see is really whether or not the process as a whole is running. In both versions of backpressure (current vs proposed here), the consensus goroutines are stopped once the buffer fills up. In the current version, they _may_ resume on their own without operator intervention. 

By performing backpressure in this way, we also lose the ability to perform the other functions of the node once the buffer has filled because the node will have crashed. In the current scheme, backpressure stalls consensus, but the other facets of the node: p2p, statesync, blocksync etc. are still available.

Reply:
Think of this model as providing a dedicated, reliable "store" for the data we're starting off sending with this ADR. Failing to send this data effectively would be the equivalent of failing to save ABCI responses [here](https://github.com/tendermint/tendermint/blob/0c96f0b434c1d9f23883d1f2291135830a7a5290/state/execution.go#L210) or failing to [save a block](https://github.com/tendermint/tendermint/blob/0c96f0b434c1d9f23883d1f2291135830a7a5290/store/store.go#L352). In either case we'd panic.

This is also why I've strongly recommended that this interface only be used from full nodes and not validators (we may even want to enforce this if we go with this solution).

The benefit of this approach over just dumping this data into a separate store on disk altogether and having the consumer read from _that_ separate store, however, is that the consumer gets real-time notifications of incoming updates, _and_ they get stronger delivery guarantees than with the subscription interface.

If there's a good database we could use to dump this data into that also provides real-time update notifications to other clients reading from that same database, then that could be a good alternative to this ADR. I've also looked into solutions like using a message broker (e.g. AMQP/Kafka/etc.), but they're all designed to handle small messages (on the order of kilobytes, not megabytes).

Reply:
Also, with this approach, if the node falls behind the rest of the network, we already have catch-up mechanisms built in to a full node. As the full node catches up, it would stream all of this data to the companion, effectively providing eventual consistency to the companion.

Reply:
> Failing to send this data effectively would be the equivalent of failing to save ABCI responses [here](https://github.com/tendermint/tendermint/blob/0c96f0b434c1d9f23883d1f2291135830a7a5290/state/execution.go#L210) or failing to [save a block](https://github.com/tendermint/tendermint/blob/0c96f0b434c1d9f23883d1f2291135830a7a5290/store/store.go#L352). In either case we'd panic.

The main difference, is that this is an external (untrusted) service. The cases you mention are within the domain of the Tendermint node - we know what we're delivering and we can make guarantees around that. I suppose it is possible to say that consensus blocks on saving a block but practically speaking this has different implications than relying on a different process (potentially on a different network).

I agree with William's conceptual framing. The mechanics of channels/buffers mean that if we wan't to fix them in size and don't want to drop data then the process must block for the channel to be emptied. It's one of those triangles where we have to choose 2 of the 3 points. I don't think we should present to users as if we have some silver bullet that does all 3. In this case, the users of this function want to not drop data and want to control how much memory grows, thus they should accept that in unideal circumstances, consensus will block.

I agree with you that failing in a loud way that lets operators know there is a problem instead of simply blocking and waiting makes more sense.

Reply:
> The main difference, is that this is an external (untrusted) service.

I'd disagree here. Would you consider a socket-based ABCI app an untrusted external service? Because the configuration for the companion service is specified in the Tendermint config in the same way as for an ABCI socket client.